File: python/querygraph/crypto.py
"""Real Ed25519 signing for QueryGraph TypeDID envelopes and lineage attestations.

Mirrors the Rust port's TypeSec `Ed25519DidKey::from_seed` pattern: a 32-byte
private key is derived deterministically from a seed via SHA-256, so agents
recreated from the same seed sign identically across processes. Signatures are
prefixed `ed25519:`; when the optional `cryptography` dependency is missing,
callers fall back to digests prefixed `unsigned:sha256:` that can never be
mistaken for signatures.

Install with `pip install querygraph[crypto]`.
"""

from __future__ import annotations

from hashlib import sha256

from querygraph.base58 import b58encode

try:  # pragma: no cover - trivially exercised by import.
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey,
        Ed25519PublicKey,
    )
    from cryptography.exceptions import InvalidSignature

    CRYPTO_AVAILABLE = True
except ImportError:  # pragma: no cover - depends on optional extra.
    CRYPTO_AVAILABLE = False

# Multicodec prefix for ed25519-pub (0xed, varint-encoded), per the did:key spec.
_ED25519_MULTICODEC = bytes([0xED, 0x01])

SIGNATURE_PREFIX = "ed25519:"
UNSIGNED_PREFIX = "unsigned:sha256:"


class CryptoUnavailableError(RuntimeError):
    def __init__(self) -> None:
        super().__init__(
            "Real Ed25519 signing requires the 'cryptography' package. "
            "Install querygraph[crypto]."
        )


def _require_crypto() -> None:
    if not CRYPTO_AVAILABLE:
        raise CryptoUnavailableError()


def _as_bytes(value: bytes | str) -> bytes:
    return value.encode() if isinstance(value, str) else value


class Ed25519Signer:
    """A deterministic Ed25519 keypair derived from a seed, or a random one."""

    def __init__(self, private_key: "Ed25519PrivateKey") -> None:
        self._private_key = private_key
        self._public_key = private_key.public_key()

    @classmethod
    def from_seed(cls, seed: bytes | str) -> "Ed25519Signer":
        _require_crypto()
        return cls(Ed25519PrivateKey.from_private_bytes(sha256(_as_bytes(seed)).digest()))

    @classmethod
    def generate(cls) -> "Ed25519Signer":
        _require_crypto()
        return cls(Ed25519PrivateKey.generate())

    @property
    def public_key_bytes(self) -> bytes:
        from cryptography.hazmat.primitives.serialization import (
            Encoding,
            PublicFormat,
        )

        return self._public_key.public_bytes(Encoding.Raw, PublicFormat.Raw)

    @property
    def public_key_multibase(self) -> str:
        return f"z{b58encode(_ED25519_MULTICODEC + self.public_key_bytes)}"

    def did_key(self) -> str:
        """The W3C did:key identifier for this keypair."""
        return f"did:key:{self.public_key_multibase}"

    def verification_method(self) -> str:
        did = self.did_key()
        return f"{did}#{self.public_key_multibase}"

    def sign(self, message: bytes | str) -> str:
        """Sign and return an `ed25519:<hex>` signature string."""
        return SIGNATURE_PREFIX + self._private_key.sign(_as_bytes(message)).hex()


def public_key_from_multibase(multibase: str) -> bytes:
    """Decode a `z…` base58btc multibase key, stripping the ed25519 multicodec."""
    from querygraph.base58 import b58decode

    if not multibase.startswith("z"):
        raise ValueError(f"Unsupported multibase prefix in {multibase!r}")
    raw = b58decode(multibase[1:])
    if raw.startswith(_ED25519_MULTICODEC):
        raw = raw[len(_ED25519_MULTICODEC) :]
    if len(raw) != 32:
        raise ValueError("Expected a 32-byte Ed25519 public key")
    return raw


def public_key_from_did_key(did: str) -> bytes:
    """Resolve a did:key (optionally with fragment) to raw public-key bytes."""
    identifier = did.split("#", 1)[0]
    if not identifier.startswith("did:key:"):
        raise ValueError(f"Not a did:key identifier: {did!r}")
    return public_key_from_multibase(identifier.removeprefix("did:key:"))


def verify(public_key: bytes | str, message: bytes | str, signature: str) -> bool:
    """Verify an `ed25519:<hex>` signature against a message.

    `public_key` may be raw bytes, a multibase string, or a did:key identifier.
    Returns False (never raises) for wrong keys, tampered messages, or
    `unsigned:` digests — those are not signatures.
    """
    _require_crypto()
    if not signature.startswith(SIGNATURE_PREFIX):
        return False
    if isinstance(public_key, str):
        try:
            public_key = (
                public_key_from_did_key(public_key)
                if public_key.startswith("did:key:")
                else public_key_from_multibase(public_key)
            )
        except ValueError:
            return False
    try:
        Ed25519PublicKey.from_public_bytes(public_key).verify(
            bytes.fromhex(signature.removeprefix(SIGNATURE_PREFIX)),
            _as_bytes(message),
        )
        return True
    except (InvalidSignature, ValueError):
        return False


def unsigned_digest(message: bytes | str) -> str:
    """A clearly-labelled non-signature digest for crypto-less environments."""
    return UNSIGNED_PREFIX + sha256(_as_bytes(message)).hexdigest()

File: typescript/src/crypto.ts
import { createHash, createPrivateKey, createPublicKey, sign as nodeSign, verify as nodeVerify } from "node:crypto";
import { b58encode, b58decode } from "./base58.js";

export const SIGNATURE_PREFIX = "ed25519:";
const PRIVATE_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const PUBLIC_PREFIX = Buffer.from("302a300506032b6570032100", "hex");
export function sha256Hex(value: string | Uint8Array): string { return createHash("sha256").update(value).digest("hex"); }
function asBytes(value: string | Uint8Array): Buffer { return typeof value === "string" ? Buffer.from(value) : Buffer.from(value); }
export class Ed25519Signer {
  private constructor(private readonly seed: Buffer, private readonly privateKey: ReturnType<typeof createPrivateKey>, private readonly publicKey: Buffer) {}
  static fromSeed(seed: string): Ed25519Signer { const bytes = createHash("sha256").update(seed).digest(); const privateKey = createPrivateKey({ key: Buffer.concat([PRIVATE_PREFIX, bytes]), format: "der", type: "pkcs8" }); const publicDer = createPublicKey(privateKey).export({ format: "der", type: "spki" }) as Buffer; return new Ed25519Signer(bytes, privateKey, publicDer.subarray(-32)); }
  sign(message: string | Uint8Array): string { return `${SIGNATURE_PREFIX}${nodeSign(null, asBytes(message), this.privateKey).toString("base64url")}`; }
  publicKeyBytes(): Buffer { return Buffer.from(this.publicKey); }
  verificationMethod(): string { return `did:key:${this.didKey()}#${this.didKey()}`; }
  didKey(): string { return `z${b58encode(Uint8Array.from([0xed, 0x01, ...this.publicKey]))}`; }
}
export function publicKeyFromDidKey(did: string): Buffer { const value = (did.split("#")[0] ?? "").replace(/^did:key:/, ""); const decoded = Buffer.from(b58decode(value.slice(1))); return decoded.subarray(2); }
export function verify(publicKey: string | Uint8Array, message: string | Uint8Array, signature: string): boolean { if (!signature.startsWith(SIGNATURE_PREFIX)) return false; const raw = typeof publicKey === "string" ? publicKeyFromDidKey(publicKey) : Buffer.from(publicKey); const key = createPublicKey({ key: Buffer.concat([PUBLIC_PREFIX, raw]), format: "der", type: "spki" }); return nodeVerify(null, asBytes(message), key, Buffer.from(signature.slice(SIGNATURE_PREFIX.length), "base64url")); }
export function unsignedDigest(message: string | Uint8Array): string { return `unsigned:sha256:${sha256Hex(message)}`; }
